leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/oscore/cf/InMemoryOscoreContextDB.java

```diff
@@ -0,0 +1,99 @@
+ * 
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v2.0
+ * and Eclipse Distribution License v1.0 which accompany this distribution.
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v20.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *     Sierra Wireless - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.core.californium.oscore.cf;
+
+import org.eclipse.californium.oscore.CoapOSException;
+import org.eclipse.californium.oscore.HashMapCtxDB;
+import org.eclipse.californium.oscore.OSCoreCtx;
+import org.eclipse.californium.oscore.OSCoreCtxDB;
+import org.eclipse.californium.oscore.OSException;
+
+/**
+ * An {@link OSCoreCtxDB} which store context in memory and is able to derive context from data provided in
+ * {@link OscoreStore}
+ *
+ */
+// TODO OSCORE this should be moved in californium.
+// TODO OSCORE I don't know if we need to extends HashMapCtxDB or implement OSCoreCtxDB
+public class InMemoryOscoreContextDB extends HashMapCtxDB {
+
+    private OscoreStore store;
+
+    public InMemoryOscoreContextDB(OscoreStore oscoreStore) {
+        this.store = oscoreStore;
+    }
+
+    @Override
+    public synchronized OSCoreCtx getContext(byte[] rid, byte[] IDContext) throws CoapOSException {
+        OSCoreCtx osCoreCtx = super.getContext(rid, IDContext);
+
+        if (osCoreCtx == null && IDContext != null) {
+            throw new IllegalArgumentException("Internal Leshan operations should always use a null ID Context");
+        }
+
+        if (osCoreCtx == null) {
+            OscoreParameters params = store.getOscoreParameters(rid);
+            // TODO OSCORE we should handle if param is null ?
+            osCoreCtx = deriveContext(params);
+            super.addContext(osCoreCtx);
+        }
+
+        return osCoreCtx;
+    }
+
+    @Override
+    public synchronized OSCoreCtx getContext(byte[] rid) {
+        OSCoreCtx osCoreCtx = super.getContext(rid);
+
+        if (osCoreCtx == null) {
+            OscoreParameters params = store.getOscoreParameters(rid);
+            // TODO OSCORE we should handle if param is null ?
+            osCoreCtx = deriveContext(params);
+            super.addContext(osCoreCtx);
+        }
+
+        return osCoreCtx;
+
+    }
+
+    @Override
+    public synchronized OSCoreCtx getContext(String uri) throws OSException {
+
+        OSCoreCtx osCoreCtx = super.getContext(uri);
+
+        if (osCoreCtx == null) {
+            byte[] rid = store.getRecipientId(uri);
+            if (rid != null) {
+                osCoreCtx = getContext(rid);
+                // TODO don't know if I should add by uri here ?
+                // super.addContext(uri, osCoreCtx);
+            }
+        }
```

Comment:
(from @rikard-sics)
> Here I would add the context to the DB associated to the URI. As here the URI would be the "foreign peer address".
>  Then this would allow the ObjectSecurityLayer to find the correct context for later outgoing requests to that peer.

Thinking a bit more about this :   
In LWM2M the foreign peer address could change (e.g. because devices are behind NAT).
When address changes client should update its registration.

So to handle this
1. either we need to never cache the context by uri  in DB (but that means more call to registration store.
2. Or we cache in for uri and we need to update/clean DB on registration update. (still a bit afraid by this kind of house keeping)

There is another solution and this is something we will maybe need anyway. 
At LWM2M level when we send a request we can tell the "identity" of the destination. This should even be done to be sure that context does not change between the moment where user decide to send the request and the moment where we really send the request.
So when we send the request we could say the expected  OSCORE foreign peer RID.
(This is what we do for DTLS see : https://github.com/eclipse/leshan/blob/leshan-2.0.0-M6/leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/request/CoapRequestBuilder.java#L304-L306)

If we do that I guess ObjectSecurityLayer can find RID in EndpointContext and so get by RID and not by URI, right ?

Reply:
Right, so you mean that the ObjectSecurityLayer would get the RID from something like the following?
`endpointContext.get(OSCoreEndpointContextInfo.OSCORE_RECIPIENT_ID)`

That could be an addition to Californium, so if that value is set the ObjectSecurityLayer would use the RID there to get the context, rather than the request URI.

Reply:
I just implemented oscore store at bs server side and this idea :point_up:  would help a lot.

Reply:
> I just implemented oscore store at bs server side and this idea point_up would help a lot.

I see, then that is something I can add to the Californium side.